crates/pg_workspace_new/src/workspace/server/change.rs

```diff
@@ -54,235 +69,278 @@ impl Document {
+                    self.move_ranges(
+                        affected_range.end(),
+                        change.diff_size(),
+                        change.is_addition(),
+                    );
 
-                let new_id = self.id_generator.next();
-                self.statements[stmt.0] = (new_id, new_stmt.add(start));
-
-                let changed_stmt = ChangedStatement {
-                    old: self.statement(&stmt.1),
-                    new_ref: self.statement_ref(&self.statements[stmt.0]),
-                    // change must be relative to statement
-                    range: change.range.unwrap().sub(stmt.1 .1.start()),
-                    text: change.text.clone(),
-                };
+                    self.content = new_content;
 
-                changed.push(StatementChange::Modified(changed_stmt));
-            } else {
-                // delete and add new ones
-                for (_, (id, r)) in &affected {
-                    changed.push(StatementChange::Deleted(self.statement_ref(&(*id, *r))));
+                    return changed;
                 }
 
-                // remove affected statements
-                self.statements
-                    .retain(|(id, _)| !affected.iter().any(|(affected_id, _)| id == affected_id));
-
-                // add new statements
-                for range in ranges {
-                    match self
-                        .statements
-                        .binary_search_by(|(_, r)| r.start().cmp(&range.start()))
-                    {
-                        Ok(_) => {}
-                        Err(pos) => {
-                            let new_id = self.id_generator.next();
-                            self.statements.insert(pos, (new_id, range));
-                            changed.push(StatementChange::Added(Statement {
-                                ref_: StatementRef {
-                                    path: self.path.clone(),
-                                    id: new_id,
-                                },
-                                text: new_content[range].to_string(),
-                            }));
-                        }
-                    }
-                }
-            }
-        }
+                let affected_idx = affected_indices[0];
+                let new_range = new_ranges[0].add(affected_range.start());
+                let (old_id, old_range) = self.positions[affected_idx];
 
-        self.content = new_content;
+                // move all statements after the afffected range
+                self.move_ranges(old_range.end(), change.diff_size(), change.is_addition());
 
-        self.debug_statements();
+                let new_id = self.id_generator.next();
+                self.positions[affected_idx] = (new_id, new_range);
 
-        changed
-    }
-}
+                changed.push(StatementChange::Modified(ModifiedStatement {
+                    old_stmt: Statement {
+                        id: old_id,
```

Comment:
at this point, that `id` is only stored here, because statement matching that id in the `document` is overwritten, right? 
Do we at some point after this try to find the old statement by id?